File: packages/server/src/domains/flag-environments/repo.ts
import { getOne, getMany } from '../../shared/db.js';
import type { FlagEnvironment, UpdateFlagEnvironmentInput } from '@flick/shared';

export async function findByFlag(flagId: string): Promise<FlagEnvironment[]> {
  return getMany<FlagEnvironment>(
    `SELECT fe.* FROM flag_environments fe
     JOIN environments e ON e.id = fe.environment_id
     WHERE fe.flag_id = $1
     ORDER BY e.sort_order ASC`,
    [flagId],
  );
}

export async function findById(id: string): Promise<FlagEnvironment | null> {
  return getOne<FlagEnvironment>('SELECT * FROM flag_environments WHERE id = $1', [id]);
}

export async function findByFlagAndEnv(flagId: string, envId: string): Promise<FlagEnvironment | null> {
  return getOne<FlagEnvironment>(
    'SELECT * FROM flag_environments WHERE flag_id = $1 AND environment_id = $2',
    [flagId, envId],
  );
}

export async function findByEnv(envId: string): Promise<FlagEnvironment[]> {
  return getMany<FlagEnvironment>(
    'SELECT * FROM flag_environments WHERE environment_id = $1',
    [envId],
  );
}

export async function update(id: string, input: UpdateFlagEnvironmentInput): Promise<FlagEnvironment | null> {
  const fields: string[] = [];
  const values: unknown[] = [];
  let idx = 1;

  if (input.enabled !== undefined) { fields.push(`enabled = $${idx++}`); values.push(input.enabled); }
  if (input.gate_config !== undefined) { fields.push(`gate_config = $${idx++}`); values.push(JSON.stringify(input.gate_config)); }

  if (fields.length === 0) return findById(id);

  values.push(id);
  return getOne<FlagEnvironment>(
    `UPDATE flag_environments SET ${fields.join(', ')} WHERE id = $${idx} RETURNING *`,
    values,
  );
}

export async function toggle(id: string): Promise<FlagEnvironment | null> {
  return getOne<FlagEnvironment>(
    'UPDATE flag_environments SET enabled = NOT enabled WHERE id = $1 RETURNING *',
    [id],
  );
}

File: packages/server-rust/src/domains/flag_environments/repo.rs
use chrono::{DateTime, Utc};
use serde::Serialize;
use serde_json::Value;
use sqlx::PgPool;
use uuid::Uuid;

#[derive(Debug, sqlx::FromRow, Serialize)]
pub struct FlagEnvironmentRow {
    pub id: Uuid,
    pub flag_id: Uuid,
    pub environment_id: Uuid,
    pub enabled: bool,
    pub gate_config: Value,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

pub async fn find_by_flag(
    db: &PgPool,
    flag_id: Uuid,
) -> Result<Vec<FlagEnvironmentRow>, sqlx::Error> {
    sqlx::query_as::<_, FlagEnvironmentRow>(
        "SELECT fe.*
         FROM flag_environments fe
         JOIN environments e ON e.id = fe.environment_id
         WHERE fe.flag_id = $1
         ORDER BY e.sort_order ASC",
    )
    .bind(flag_id)
    .fetch_all(db)
    .await
}

pub async fn find_by_id(
    db: &PgPool,
    id: Uuid,
) -> Result<Option<FlagEnvironmentRow>, sqlx::Error> {
    sqlx::query_as::<_, FlagEnvironmentRow>(
        "SELECT * FROM flag_environments WHERE id = $1",
    )
    .bind(id)
    .fetch_optional(db)
    .await
}

pub async fn find_by_flag_and_env(
    db: &PgPool,
    flag_id: Uuid,
    env_id: Uuid,
) -> Result<Option<FlagEnvironmentRow>, sqlx::Error> {
    sqlx::query_as::<_, FlagEnvironmentRow>(
        "SELECT * FROM flag_environments WHERE flag_id = $1 AND environment_id = $2",
    )
    .bind(flag_id)
    .bind(env_id)
    .fetch_optional(db)
    .await
}

pub async fn find_by_env(
    db: &PgPool,
    env_id: Uuid,
) -> Result<Vec<FlagEnvironmentRow>, sqlx::Error> {
    sqlx::query_as::<_, FlagEnvironmentRow>(
        "SELECT * FROM flag_environments WHERE environment_id = $1",
    )
    .bind(env_id)
    .fetch_all(db)
    .await
}

pub async fn update(
    db: &PgPool,
    id: Uuid,
    enabled: Option<bool>,
    gate_config: Option<Value>,
) -> Result<Option<FlagEnvironmentRow>, sqlx::Error> {
    if enabled.is_none() && gate_config.is_none() {
        return find_by_id(db, id).await;
    }

    let mut query = String::from("UPDATE flag_environments SET updated_at = NOW()");
    let mut param_idx = 0u32;

    if enabled.is_some() {
        param_idx += 1;
        query.push_str(&format!(", enabled = ${param_idx}"));
    }

    if gate_config.is_some() {
        param_idx += 1;
        query.push_str(&format!(", gate_config = ${param_idx}"));
    }

    param_idx += 1;
    query.push_str(&format!(" WHERE id = ${param_idx} RETURNING *"));

    let mut q = sqlx::query_as::<_, FlagEnvironmentRow>(&query);

    if let Some(e) = enabled {
        q = q.bind(e);
    }

    if let Some(gc) = gate_config {
        q = q.bind(gc);
    }

    q = q.bind(id);

    q.fetch_optional(db).await
}

pub async fn toggle(
    db: &PgPool,
    id: Uuid,
) -> Result<Option<FlagEnvironmentRow>, sqlx::Error> {
    sqlx::query_as::<_, FlagEnvironmentRow>(
        "UPDATE flag_environments SET enabled = NOT enabled, updated_at = NOW()
         WHERE id = $1 RETURNING *",
    )
    .bind(id)
    .fetch_optional(db)
    .await
}
